pkg/client/tcp_client.go

```diff
@@ -1,28 +1,224 @@
 package client
 
 import (
+	"bufio"
 	"fmt"
+	"io"
 	"net"
+
+	"github.com/langorou/langorou/pkg/utils"
+)
+
+// A good ressource
+// https://ipfs.io/ipfs/QmfYeDhGH9bZzihBUDEQbCbTc5k5FZKURMUoUvfmc27BwL/socket/tcp_sockets.html
+
+// About net.Conn vs net.TCPConn :
+// > This interface has primary methods ReadFrom and WriteTo to handle packet reads and writes.
+// > The Go net package recommends using these interface types rather than the concrete ones.
+// > But by using them, you lose specific methods such as SetKeepAlive or TCPConn and SetReadBuffer of UDPConn, unless you do a type cast. It is your choice.
+
+// TCPClient connects to the game server
+
+type ClientMsg int
+
+const (
+	NME ClientMsg = iota
+	MOV
+)
+
+type ServerCmd int
+
+const (
+	_ ServerCmd = iota
+	SET
+	HUM
+	HME
+	MAP
+	UPD
+	END
+	BYE
 )
 
-type TcpClient struct {
-	conn *net.Conn
+func (cmd ServerCmd) String() string {
+	return [...]string{"UNKNOWN", "SET", "HUM", "HME", "MAP", "UPD", "END", "BYE"}[cmd]
 }
 
-// NewTcpClient creates a new TCP client
-func NewTcpClient(ipAddr string, port string) (*TcpClient, error) {
+type TCPClient struct {
+	conn net.Conn
+}
 
-	ip := net.ParseIP(ipAddr)
-	if ip == nil {
-		return nil, fmt.Errorf("invalid IP address: %s", ipAddr)
-	}
+// NewTCPClient creates a new TCP client
+func NewTCPClient(addr string) (TCPClient, error) {
 
-	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%s", ipAddr, port))
+	// We might need to use net.DialTCP with https://golang.org/pkg/net/#TCPConn.SetKeepAlive
+	conn, err := net.Dial("tcp", addr)
 	if err != nil {
-		return nil, err
+		return TCPClient{}, err
 	}
 
-	return &TcpClient{
-		conn: &conn,
+	return TCPClient{
+		conn: conn,
 	}, nil
 }
+
+// SendName to the server
+func (c TCPClient) SendName(name string) error {
+	isASCII := utils.IsASCII(name)
+	if !isASCII {
+		return fmt.Errorf("%s is not a valid ASCII name", name)
+	}
+
+	t := len(name)
+	if t == 0 || t > 255 {
+		return fmt.Errorf("invalid name '%s', please use a short ASCII name", name)
+	}
+
+	msg := make([]byte, 3+1+t)
+	fmt.Println(3 + 1 + t)
+	copy(msg[:3], []byte("NME"))
+	msg[3] = byte(uint8(t))
+	copy(msg[4:], []byte(name))
+
+	_, err := c.conn.Write(msg)
+
+	return err
+}
+
+// SendMove to the server
+func (c TCPClient) SendMove(n uint8, moves []Move) error {
+
+	msg := make([]byte, 3+1+5*n)
+
+	copy(msg[:3], []byte("MOV"))
+	msg[3] = byte(n)
+
+	for i := 0; i < int(n); i++ {
+		msg[4+5*i] = byte(moves[i].Start.X)
+		msg[4+5*i+1] = byte(moves[i].Start.Y)
+		msg[4+5*i+2] = byte(moves[i].N)
+		msg[4+5*i+3] = byte(moves[i].End.X)
+		msg[4+5*i+4] = byte(moves[i].End.Y)
+	}
+
+	_, err := c.conn.Write(msg)
+
+	return err
+}
+
+// ReceiveMsg from the server and parse it
+func (c TCPClient) ReceiveMsg() (ServerCmd, error) {
```

Comment:
```suggestion
func (c *TCPClient) ReceiveMsg() (ServerCmd, error) {
```

we will want to use a pointer here (otherwise the modifications done on the client state might not be persisted across messages)